pkg/ring/kv/memberlist/memberlist_client.go

```diff
@@ -475,6 +493,33 @@ func (m *KV) joinMembersOnStartup(ctx context.Context, members []string) error {
 	return lastErr
 }
 
+// Provides a dns-based member disovery to join a memberlist cluster w/o knowning members' addresses upfront.
+func (m *KV) discoverMembers(ctx context.Context, members []string) []string {
+	if len(members) == 0 {
+		return nil
+	}
+
+	var ms, resolve []string
+
+	for _, member := range members {
+		if strings.Contains(member, "+") {
+			resolve = append(resolve, member)
+		} else {
+			// No DNS SRV record to lookup, just append member
+			ms = append(ms, member)
+		}
+	}
+
+	err := m.provider.Resolve(ctx, resolve)
+	if err != nil {
+		level.Error(util.Logger).Log("msg", "failed to resolve members from SRV records", "addrs", strings.Join(resolve, ","))
```

Comment:
I'd drop `from SRV records` part from the message, resolver may support different ways of resolving.